### PR TITLE
[KYUUBI #1143] Refact Serverable/FrontendService/ServiceDiscovery hierarchy

### PR DESCRIPTION
**Pull request by yaooqinn**
### _Why are the changes needed?_

After this PR, the service named service discovery will be a child service of frontend service and it's optional.

A servalable will contain a backend and a list of frontends, the former one stay AS-IS and the latter is changed from single to multiple, which means a serverable can have multiple interfaces to interact with its different clients

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @turboFei @yanghua @cfmcgrady @ulysses-you @pan3793 thanks

File: kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala [hunk 1]
```diff
@@ -55,7 +55,7 @@ case class KyuubiSessionEvent(
 object KyuubiSessionEvent {
   def apply(session: KyuubiSessionImpl): KyuubiSessionEvent = {
     assert(KyuubiServer.kyuubiServer != null)
-    val serverIP = KyuubiServer.kyuubiServer.connectionUrl
+    val serverIP = KyuubiServer.kyuubiServer.frontendServices.head.connectionUrl
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala:58]: this is quite hacky, but not related to this change

Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala:58]: I think we can add a method in `Utils` which check `frontendServices` only has one element and extract `connectionUrl` from the it.

Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala:58]: that's hacky too

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/SparkThriftBinaryFrontendServiceSuite.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      conf.set(KyuubiConf.ENGINE_CONNECTION_URL_USE_HOSTNAME, true)
+      service.initialize(conf)
+      // default use hostname
+      assert(service.connectionUrl.startsWith("localhost"))
+
+      // use ip address
+      conf.set(KyuubiConf.ENGINE_CONNECTION_URL_USE_HOSTNAME, false)
+      assert(service.connectionUrl.startsWith("127.0.0.1"))
+    } finally {
+      conf.set(KyuubiConf.ENGINE_CONNECTION_URL_USE_HOSTNAME, true)
```
Comment (yanghua) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/SparkThriftBinaryFrontendServiceSuite.scala:40]: Is it necessary for this `finally` block?

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/SparkThriftBinaryFrontendServiceSuite.scala:40]: seems not

**Review by ulysses-you — APPROVED** [review]
So now the relationship looks like:
```
    server
   /   |   \
  fe   fe   fe ...
  |    |    |
  be   be   be ...
```

looks good, we can add other frontend more easy in futrue.

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/SparkThriftBinaryFrontendServiceSuite.scala [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+package org.apache.kyuubi.engine.spark
+
+import org.apache.kyuubi.KyuubiFunSuite
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.service.NoopServer
+
+class SparkThriftBinaryFrontendServiceSuite extends KyuubiFunSuite {
+  test("engine connect url use hostname") {
```
Comment (cfmcgrady) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/SparkThriftBinaryFrontendServiceSuite.scala:25]: It seems the unit test can't pass on my laptop.

```scala
"192.168.207.81:10009" did not start with "localhost"
ScalaTestFailureLocation: org.apache.kyuubi.engine.spark.SparkThriftBinaryFrontendServiceSuite at (SparkThriftBinaryFrontendServiceSuite.scala:33)
org.scalatest.exceptions.TestFailedException: "192.168.207.81:10009" did not start with "localhost"
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
	at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
	at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
	at org.apache.kyuubi.engine.spark.SparkThriftBinaryFrontendServiceSuite.$anonfun$new$1(SparkThriftBinaryFrontendServiceSuite.scala:33)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
	at org.apache.kyuubi.KyuubiFunSuite.withFixture(KyuubiFunSuite.scala:55)
	at org.apache.kyuubi.KyuubiFunSuite.withFixture$(KyuubiFunSuite.scala:49)
	at org.apache.kyuubi.engine.spark.SparkThriftBinaryFrontendServiceSuite.withFixture(SparkThriftBinaryFrontendServiceSuite.scala:24)
	at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:224)
	at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:236)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:236)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:218)
	at org.apache.kyuubi.engine.spark.SparkThriftBinaryFrontendServiceSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkThriftBinaryFrontendServiceSuite.scala:24)
	at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:234)
	at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:227)
	at org.apache.kyuubi.engine.spark.SparkThriftBinaryFrontendServiceSuite.runTest(SparkThriftBinaryFrontendServiceSuite.scala:24)
	at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTests$1(AnyFunSuiteLike.scala:269)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTests(AnyFunSuiteLike.scala:269)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTests$(AnyFunSuiteLike.scala:268)
	at org.scalatest.funsuite.AnyFunSuite.runTests(AnyFunSuite.scala:1563)
	at org.scalatest.Suite.run(Suite.scala:1112)
	at org.scalatest.Suite.run$(Suite.scala:1094)
	at org.scalatest.funsuite.AnyFunSuite.org$scalatest$funsuite$AnyFunSuiteLike$$super$run(AnyFunSuite.scala:1563)
	at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$run$1(AnyFunSuiteLike.scala:273)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
	at org.scalatest.funsuite.AnyFunSuiteLike.run(AnyFunSuiteLike.scala:273)
	at org.scalatest.funsuite.AnyFunSuiteLike.run$(AnyFunSuiteLike.scala:272)
	at org.apache.kyuubi.engine.spark.SparkThriftBinaryFrontendServiceSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkThriftBinaryFrontendServiceSuite.scala:24)
	at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
	at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
	at org.apache.kyuubi.engine.spark.SparkThriftBinaryFrontendServiceSuite.run(SparkThriftBinaryFrontendServiceSuite.scala:24)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13(Runner.scala:1322)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13$adapted(Runner.scala:1316)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1316)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24(Runner.scala:993)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24$adapted(Runner.scala:971)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1482)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:971)
	at org.scalatest.tools.Runner$.run(Runner.scala:798)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:133)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:27)
```

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/SparkThriftBinaryFrontendServiceSuite.scala:26]: right, fixed

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**